### ISSUE TITLE
symmetric?

### ISSUE THREAD

**Issue by mars0i** [mikera/core.matrix#92]
I wanted a test for a matrix being symmetric, and wrote one.  Then thought it might be useful to have in core.matrix, added it to core.matrix and vectorz-clj, and rewrote it to speed it up.  Does it seem worthwhile to include a `symmetric` predicate in core.matrix?

The changes can be found in files 
src/main/clojure/clojure/core/matrix/impl/default.clj
src/main/clojure/clojure/core/matrix/protocols.clj
src/main/clojure/clojure/core/matrix.clj
at https://github.com/mars0i/core.matrix
and
src/main/clojure/mikera/vectorz/matrix_api.clj
at https://github.com/mars0i/vectorz-clj .

I'm pretty new to Clojure, and to collaborating on others' projects, so I'm not sure that I went about this the right way.  The logical place to put `symmetric?` seemed to be in the `PMatrixPredicates` protocol, and that's what I did.  However, `PMatrixPredicates` is shadowed in vectorz-clj, so I had to add the test separately there.  I tried to emulate some code I found in vectorz-clj and vectorz, but mainly just pasted the same code into vectorz-clj.  I imagine this should be reimplemented in vectorz at some point.  (I'm willing to try to do this if it seems worthwhile.  Should be pretty easy, even though I'm pretty rusty with Java, but I might get tripped up on integrating jars or something like that.)

The significant part of the code in core.matrix looks like this:

``` clojure
;; Helper function for symmetric? predicate in PMatrixPredicates.
;; Note loop/recur instead of letfn/recur is 20-25% slower.
(defn- symmetric-matrix-entries?
  [m]
  (let [dim (first (mp/get-shape m))]
    (letfn [(f [i j]
              (cond
                (>= i dim) true ; all entries match: symmetric
                (>= j dim) (recur (+ 1 i) (+ 2 i)) ; all j's OK: restart with new i
                (= (mp/get-2d m i j)
                   (mp/get-2d m j i)) (recur i (inc j)) ; OK, so check next pair
                :else false))] ; not same, not symmetric
      (f 0 1))))
```

I separated out this function from the main routine in PMatrixPredicates because, at least for now, it could be called from vectorz-clj.  However, I'm not sure where to put `symmetric-matrix-entries?` for this purpose.  All of the source files that seemed like appropriate places for it would have required changing parts of `ns` directives, and I didn't want to do that.  So for now, `symmetric-matrix-entries?` appears in both src/main/clojure/clojure/core/matrix/impl/default.clj and src/main/clojure/mikera/vectorz/matrix_api.clj .

This is a simple addition, and someone else who knows more may well have a better way to implement the same function.  I've tested it with persistent-vector, ndarray, vectorz, and clatrix in Clojure, and have run different versions through criterium to see what was fastest.  (Oddly, changing the same code to use `loop`/`recur` is 20-25% makes it slower on moderately sized matrices, e.g. 250x250 to 1000x1000.  I haven't tested larger matrices.)   Are there other tests I should run?

**Comment by mikera** [comment]
`symmetric?` is certainly a useful predicate, I'd be happy to take patches to implement this.

Your approach sounds like it is generally in the right direction.

Some implementation points:
- `PMatrixPredicates` would be a sensible place to put the protocol. There is a bit of trickiness around updating all the implementations, but I think this is manageable
- `symmetric-matrix-entries?` should get called from default.clj - so sounds like you got this part right!
- `vectorz-clj` should use the `.isSymmetric()` Java method (this already exists in the Vectorz AMatrix class). This is a specialised implementation that should be much faster than a generic approach, e.g. it knows that a diagonal matrix is always symmetric etc. So `symmetric-matrix-entries?` should not be necessary for Vectorz.
- `loop/recur` should be faster.... you need to make sure all the type hints / use of primitive values are correct however. As soon as Clojure starts boxing values, tight loops over all elements start to get a bit expensive.....

**Comment by mars0i** [comment]
Great.  My implementation of `symmetric?` was slower for vectorz matrices.  `.isSymmetric` should definitely be faster.  I have a further question, but I ask it as a vectorz-clj issue.

I don't know much about boxing and type hints in Clojure.  I'll learn more and work on that.

**Comment by mars0i** [comment]
I implemented the 2D test with `.isSymmetric` for vectorz.  Incredibly fast.

I still don't understand why definining `symmetric?` with `loop` is slower than with `letfn` in Clojure 1.5.1.  Mike, what sort of type info could help?  The indexes begin as constants, so presumably the compiler knows what they are. (Tried annotating them anyway, and it didn't help.)

Seems to me that the only thing that type information could affect, then, is `get-2d`, and I could only specify matrix type information if I wrote different versions of `symmetric?` for different implementations.

That's worth doing, but `PMatrixPredicates` currently only exists in core.matrix's default.clj and vectorz-clj's matrix_api.clj.  I'm reluctant to add other implementations at this point, since I don't yet understand all of the types within each implementation, and ndarray's handy `magic` macros will require some study.  If it's useful for me to add a default implementation of `symmetric?` along with the (trivial) vectorz-clj 2D implementation, that would at least be a start.

So why don't I just use `letfn` for a default version of `symmetric?` for 2D in default.clj.  It's only very slightly more complicated-looking than the same thing using `loop`.

**Comment by mars0i** [comment]
(1) I should say up front I won't be able to spend much time on implementing `symmetric?`, unfortunately.  (Mike, as I've said, I only started using Clojure recently, and if you decide at some point that the cost to you of me trying to implement this is greater than the benefits, I won't be offended if you tell me to forget it.  I learned things in the process.)

(2) Mike A. and Konrad Hinsen (https://groups.google.com/d/msg/numerical-clojure/wUIUTwjiCig/F5s1fIotbooJ) have proposed that symmetry for N-dimensional arrays should be such be defined so that 

> `(symmetric? A)` if and only if `(= A (transpose A))`.

I won't discuss the easy case of 1D structures and scalars.  N > 2 is the interesting case.

I haven't yet fully understood the definition of `transpose` in core.matrix/default.clj yet, but from experiments it appears that this proposal would mean that one can test for symmetry by comparing each element of an array with the element picked out by reversing the order of indexes that pick out the original element:

> `(symmetric? A)` 
> if and only if 
> `(= (get-nd A i1 i2 i3 ... iN) (get-nd A iN ... i3 i2 i1))`

Is this correct?

(3) I've gotten two answers to my question about generalizations of symmetry for more than two dimensions at math.stackexchange (http://math.stackexchange.com/questions/615119/how-to-generalize-symmetry-for-higher-dimensional-arrays/619769?noredirect=1#619769). The core idea is that in testing for symmetry, you're comparing each element with an element chosen by permuting the indexes that pick out the first element.  Shuchang's answer indicated that there's no priveleged permutation; you can define different symmetry concepts by different permutations.  Calle's answer was that an N-D array could be considered symmetric if and only if each element is equal to elements at locations indexed by all possible permutations of its indexes.

(4) It's not hard to define `symmetric?` for 2-D, but it seems a bit tricky for arbitrary dimensions > 2.  I've been thinking about it, and want to give it a shot (might have to back out if it becomes too challenging for me given time constraints).

If we just pick one particular permutation of indexes to define `symmetric?`, even the one that corresponds to `transpose`, writing an N-D version seems like a lot of work to go through when there's no particular use case yet.  In fact it seems plausible given the answers at math.SE, that actual use cases could need a completely different symmetry concept than the one that corresponds to `transpose`.

(5) What I suggest is that `symmetric?` for dimensions N > 2  take an optional argument that's a collection of sequences of indexes from 0 to N.  These sequences represent permutations of indexes.  If you want only one permutation, then you pass a collection containing only one such sequence.  If you want to test for all-way symmetry in 3-D, for example, as suggested by Calle's answer, you could pass squences representing all of the permutations except `[0 1 2]` to `symmetric?`:

``` Clojure
([0 2 1] [1 0 2] [1 2 0] [2 0 1] [2 1 0])
```

Without the optional argument, `symmetric?` could default to one permutation, probably the reversal of indexes that corresponds to `transpose`.

Letting `symmetric?` deal with arbitrary permutations of indexes makes it more complex and harder to understand, but to me the more difficult problem is defining the function for arbitrary dimensions, with different numbers of indexes internally.  I think there might be a way of doing this that won't be difficult to understand. My guess is that adding code to deal with arbitrary permutations won't make the code for `symmetric?` a lot harder to understand.  Not sure how it will affect efficiency, though.  Would think about that after getting a working version.

An alternative to passing collections of sequences representing permutations would be to allow users to pass in a function that can map any sequence of indexes to a collection of sequences of indexes.  This generalizes the set-of-permutations-idea, since you could write a function that chooses different permutations for different parts of an array.  This additional capability doesn't seem worthwhile to me---it's too far from the basic concept of symmetry.  If you need some wierd concept of symmetry, you're probably going to have to do other wierd things and write your own code to do it.  Still maybe the code for `symmetric?` would be a little cleaner with a passed function rather than a collection of sequences?  Nah.  And passing the collection would be easier for users to understand.

(If I can get a default `symmetric?` working for arbitrary dimensions, I might be willing to try adding a version to vectorz, if I can use a similar algorithm in Java.  I'm not aware that there's one there now.  Not sure, though.  I considered myself a Java programmer at one time, but gave up my Java union card about ten years ago.  I write Java now and then, but no longer have Java's many routine anti-gotcha rules ingrained in me.)

Ah, posted this and then realized that some of it would be fruitful to put on the Google group.  Will do that too.

**Comment by mikera** [comment]
(1) That's fine! we all contribute what we can, this is open source after all :-) My only request is that you let us all know what you can/can't do so we can plan around it / fill in the gaps etc. If it helps, it can be useful to do smaller incremental changes as separate pull requests, which is often better than trying to do everything in one go.
(2) Your interpretation is correct
(3) This is actually a very interesting idea. I quite like the idea of checking equality across all permutations (though this may reflect a bias of mathematicians, who like to think in terms of set theory and treat all permutations equally....)
(4-5) If it is too complex to implement then I would say don't worry about the 3+ case for now. Nobody as far as I know has a proper use case for it - it's more of a theoretical curiosity. The important thing is to make 0,1 and 2 dimensions work. It's fine if we make 3+ dimensions throw UnsupportedOperationException for the moment.

**Comment by mikera** [comment]
On the default.clj version : I'd say just get a version working with decent tests and we can merge that in. 

We can optimise the performance later. There are lots of different approaches that might be effective, e.g.
- Test 1-10 randomly chosen pairs of symmetric indexes first using get-2d : only if these are equal then proceed to do an exhaustive test (this would provide quick bail-out in the case of matrices that are non-symmetric in the majority of positions)
- Use `to-double-array` to extract matrix element values in bulk, and use a quick array scanning technique to test for symmetry.
- Compare the rows of a matrix with its columns (accessed using `slice`) 
- Some combination of the above

No idea which of these would be useful (if any), but the point is to get a baseline version working and then we can experiment with these kind of improvements.

**Comment by mars0i** [comment]
This all sounds good.  Where should tests go?  clojure.core.matrix.test-api?

**Comment by mikera** [comment]
I will add some more comments to each of the test namespaces to more accurately describe what should go there. But yes, `test-api` is the generic place for putting a few tests of each API function.

The other test namespaces have more specific purposes, e.g. testing out all the quirks and assumptions of a specific implementation e.g. `test-double-array.clj` is intended to test out the behaviour and expectations of the `:double-array` implementation

**Comment by mars0i** [comment]
I think what I've got for `symmetric?` looks OK now.  Let me know what you think, and I can do a pull request if it seems ready.  The changes are in the develop repos at https://github.com/mars0i/core.matrix and https://github.com/mars0i/vectorz-clj .  (I think I did the git stuff properly--just pulled the latest develop versions at github/mikera into my repos and committed to github/mars0i, so it should be up to date.)  I'm happy to change anything, or if it's easier to pull and change yourself, that's fine, of course.

The modified files in core.matrix are:
src/main/clojure/clojure/core/matrix/impl/default.clj
src/main/clojure/clojure/core/matrix/protocols.clj
src/main/clojure/clojure/core/matrix.clj
src/test/clojure/clojure/core/matrix/test_api.clj

The modified file in vectorz-clj is:
src/main/clojure/mikera/vectorz/matrix_api.clj

Searching for "symmetric" will find the relevant parts.

Notes:

There are two implementations: a default implementation and a vectorz implementation.

For any `x` with fewer than two dimensions, `(symmetric? x)` ==> `true`.  That includes `nil`.  Or should `(symmetric? nil)` be false?  nil's shape is `nil`, so maybe that would be better.  Or return `nil`?

For `x` with more than two dimensions, `UnsupportedOperationException` is thrown.

I put the tests in `test-predicates`.  Not sure if that's best.  I put in quite a few tests--more than needed?  I like to see all the cases.  I included a test for a 3-D array, catching the exception and returning true, since that's the expected behavior.  Can take that out if it seems unnecessary or distracting.  (clatrix throws the same exception, so it catches that, too.)
